### ISSUE TITLE
More flexible panel/dialog construction

### ISSUE THREAD

**Issue by michael** [substance/substance#302]
Been thinking about better managing panels and dialogs today. In Lens we have the problem that we can't hide the context toggles when the current panel is a dialog and make it use up all space. We can't do this atm without breaking modularity (e.g. with a CSS override). The problem is that when we render the contextToggles we don't know about what panel we are going to render (a regular panel or a dialog) so this information needs to made made available where the ContextToggles are created.

Let's say we provide only the contents of our panels as component classes so we have Toc instead of TocPanel. Then we will provide a ScrollablePanel and Dialog implementation as described in #274. 

``` js
// to be wrapped into ContentPanel
ArticleBody.Prototype = function() {
  this.render = function() {
    return $$('div').append(
      var config = this.context.config;
      $$(CoverEditor).ref('coverEditor'),

      // The full fledged document (ContainerEditor)
      $$("div").ref('main').addClass('document-content').append(
        $$(ContainerEditor, {
          name: 'main',
          containerId: config.containerId,
          commands: config.main.commands,
          textTypes: config.main.textTypes
        }).ref('mainEditor')
      ),
      $$(BibliographyComponent).ref('bib')
    )
  };
};

// To be wrapped into Panel with native scrollbars
TOC.Prototype = function() {
  this.render = function() {
    // .sc-toc
    //   .se-toc-items
    //      .se-toc-item
    //      .se-toc-item
    // etc.
  };
};

// To be wrapped into a DialogPanel
AddBibItems.Prototype = function() {
  this.render = function() {
    return $$('div').append(
      // .sc-add-bib-items
      // etc.
    )
  };
};
```

Let's also define an explicit panelconfiguration on app level. That's important to have things extensible. Once we have our Package API, Lens would define an API to announce panels to the app.

``` js
var panelConfig = {
  'bib-items': {
    type: 'panel',
    props: {
      scrollbar: {
        position: 'right',
        type: 'substance'
      }
    }
  },
  'toc': {
    type: 'panel',
    props: {
      scrollbar: {
        position: 'right',
        type: 'native'
      }
    }
  },
  'add-bib-items': {
    type: 'dialog',
    props: {
      scrollbar: {
        position: 'right',
        type: 'native'
      }      
    }
  },
  ...
};
```

The current implementation for rendering contextPanel is hacky and not reusable (see https://github.com/substance/lens/blob/master/LensController.js#L108-L115 and https://github.com/substance/lens/blob/master/LensController.js#L129-L136 ):

I would provide a sidebar component (_better name it needs_) that can be used by Lens-ish 2-column apps. In this sketch our app exposes this.context.activePanel to have all info to build the sidebar according to the panelConfig we provided.

``` js
Sidebar.Prototype = function() {
  render: function() {
    // LensWriter exposes this as an API
    var activePanel = this.context.activePanel;
    var el = $$('div').addClass('sc-sidebar');
    // Only render context toggles when we are dealing with a dialog
    if (activePanel.type === 'dialog') {
      el.append($$(ContextToggles));
      el.append(
        $$(DialogPanel, activePanel.panelProps).append(
          // activePanel.content props has the doc and the the state values
          // e.g. {doc: ..., bibItemId: '12'}
          $$(activePanel.PanelContentClass, activePanel.contentProps)
        )
      )
    } else {
      // Wrap panel content into a ScrollabelPanel
      el.append(
        $$(ScrollablePanel).append(
          $$(activePanel.PanelContentClass, activePanel.props)
        )
      );
    }
    return el;
  }
};
```

Oliver does that make sense to you? I would want to solve this soon, because the display of LensWriter suffers from this limitation. Should be able to do in a few hours if confirmed. So maybe I can accomplish this within the beta2 release.

**Comment by obuchtala** [comment]
I understand and like your proposal. While it is solving this problem, it would increase code reuse even more, as the Panel content implementations are so much simplified.
go for it :+1:

**Comment by michael** [comment]
Regarding terminology... let's call the more top-level parts Sections not to be confused by Panels which are wrappers for scrollable content. Or do you have a better word for 'panel' even?

Here's two possible structures:

``` js
LensWriter
  MainSection
   Panel
     ArticleBody
  ContextSection
    ContextToggles
    Panel
      Toc
```

``` js
LensWriter
  MainSection
   Panel
     ArticleBody
  ContextSection
    Dialog
      AddBibItems
```

I'm just not sure if the wrapping component Panel we call it now is actually something less top-level. More a ScrollableElement ? ScrollableWidget... or is it just a Window? ;) Container would also fit but that would lead to confusion with our DocumentContainer class.

**Comment by obuchtala** [comment]
In Java Swing you can nest Panels, so they are not necessarily top level. There is e.g. a ScrollablePanel.
Dunno. Sorry.

**Comment by michael** [comment]
Things is that we don't do inheritance anymore where Panel is an abstract class.. From now on Panels will be instantiated wrapping what you pass as children via append. I just leave the name panel for now.. but that needs a bit of thinking I guess.

**Comment by michael** [comment]
There's a problem with this decoupled approach :(. Often the implementation wants to interact with the Panel API. E.g. in BibItemsPanel we do this:

``` js
  this.didMount = function() {
    var bibItemId = this.getFirstActiveBibItemId();
    if (bibItemId) {
      this.scrollToNode(bibItemId);
    }
  };
```

Now I would need use the sendAction API to interact with the panel that wraps me. Feels a bit weird to me. I mean I'm just realizing we have some hard dependencies btw the panel content functionality and the panel logic. It would be desirable to keep the content implementation dumb, which i hoped for using this new approach. But now it would loose it's beauti when we need to add a communication layer. What's your opinion?

**Comment by obuchtala** [comment]
Hmmm... You could use the context, alternatively.

**Comment by michael** [comment]
Maybe it's an option to not make the panel wrapping automatically but let the implementer do this. Then BibItemsPanel would use a panel.. instead of being wrapped by a panel.

``` js
TOC.prototype.render = function() {
    var el = $$('div').addClass('sc-toc').append(
      $$(Panel).ref('panelEl').append(
        tocEntries
      )
    );
    return el;
}
```

Again then we have the tight coupling but with the scrollToStuff we probably need it anyway somehow. However now the initial scrollTo can be triggered via `this.refs.panelEl`.

``` js
TOC.prototype.render = function() {
  this.didMount = function() {
    var bibItemId = this.getFirstActiveBibItemId();
    if (bibItemId) {
      this.refs.panelEl.scrollToNode(bibItemId);
    }
  };
}
```

**Comment by michael** [comment]
How you mean with context?

**Comment by obuchtala** [comment]
Panel#getChildContext() -> panel: this

**Comment by obuchtala** [comment]
Don't know... maybe I didn't understand you fully.

**Comment by obuchtala** [comment]
Refs won't work I think, because they are set for the owner.

**Comment by obuchtala** [comment]
Could you explain again why making TOC render the panel and the content is a solution?
That is the way we did it, right?

**Comment by michael** [comment]
Ah yeah context would work too.. the thing is as soon as we have a dependency to the panel.. we don't win much with separating the panel from the actual TOC component etc.

This approach would be easy to follow I think:
https://github.com/substance/substance/issues/302#issuecomment-160397917

**Comment by michael** [comment]
No the way we did it was that we extend Panel to a TocPanel and then have a convention for the panel structure.

**Comment by obuchtala** [comment]
Yeah, just to hide strong coupling with DI is rather bad.

**Comment by michael** [comment]
Which was error prone. E.g. when forget to set `se-panel-content-inner`.

**Comment by obuchtala** [comment]
Ok. So Panel is now just a better to use Component.
I think this approach is clean.

**Comment by michael** [comment]
the approch with using panel inside Toc et al would also allow to control the markup for a dialog panel.. i wouldn't want to generalize this too much.. soemtimes you want to make your own ui control is that dialog header. but if we instantiate the dialog automatically users cant't intervent. Writing an example...

**Comment by michael** [comment]
``` js
CiteDialog.prototype.render = function() {
    var el = $$('div').addClass('sc-toc').append(
      $$(CiteDialogHeader), // with back button (delete button etc.)
      // scrollable content goes here
      $$(Panel).ref('panelEl').append(
        tocEntries
      )
    );
    return el;
}
```

_cite-dialog.scss could then control the layout.

``` scss
  .sc-cite-dialog {
    .sc-header {
      position: absolute;
      top: 0px;
      left: 0px;
      right: 0px;
      height: 40px;
    }
    .sc-panel {
      position: absolute;
      top: 40px
      left: 0px;
      right: 0px;
      bottom: 0px;
    }
  }
```

**Comment by obuchtala** [comment]
Ok.

**Comment by michael** [comment]
Seems to come together quite well... finally... I've been running in cycles here for way too long. Soon to push results.

**Comment by michael** [comment]
Fixed via #308
